JetStream: possible panic on consumer delete during server shutdown

This would produce the following stack:
```
panic: close of closed channel

goroutine 2360 [running]:
github.com/nats-io/nats-server/v2/server.(*Consumer).stop(0xc0003c4e00, 0x10101, 0x0, 0x0)
	/Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/consumer.go:1722 +0xf5
github.com/nats-io/nats-server/v2/server.(*Consumer).Delete(...)
	/Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/consumer.go:1706
```

--- a/server/consumer.go
+++ b/server/consumer.go
@@ -193,6 +193,7 @@ type Consumer struct {
 	ackEventT         string
 	deliveryExcEventT string
 	created           time.Time
+	closed            bool
 }
 
 const (
@@ -1707,11 +1708,11 @@ func (o *Consumer) Delete() error {
 
 func (o *Consumer) stop(dflag, doSignal, advisory bool) error {
 	o.mu.Lock()
-	mset := o.mset
-	if mset == nil {
+	if o.closed {
 		o.mu.Unlock()
 		return nil
 	}
+	o.closed = true
 
 	if dflag && advisory {
 		o.sendDeleteAdvisoryLocked()
@@ -1721,6 +1722,7 @@ func (o *Consumer) stop(dflag, doSignal, advisory bool) error {
 	close(o.qch)
 
 	store := o.store
+	mset := o.mset
 	o.mset = nil
 	o.active = false
 	ackSub := o.ackSub

--- a/test/jetstream_test.go
+++ b/test/jetstream_test.go
@@ -584,6 +584,56 @@ func TestJetStreamPullConsumerDelayedFirstPullWithReplayOriginal(t *testing.T) {
 	}
 }
 
+func TestJetStreamNoPanicOnRaceBetweenShutdownAndConsumerDelete(t *testing.T) {
+	cases := []struct {
+		name    string
+		mconfig *server.StreamConfig
+	}{
+		{"MemoryStore", &server.StreamConfig{Name: "MY_STREAM", Storage: server.MemoryStorage}},
+		{"FileStore", &server.StreamConfig{Name: "MY_STREAM", Storage: server.FileStorage}},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			s := RunBasicJetStreamServer()
+			defer s.Shutdown()
+
+			if config := s.JetStreamConfig(); config != nil {
+				defer os.RemoveAll(config.StoreDir)
+			}
+
+			mset, err := s.GlobalAccount().AddStream(c.mconfig)
+			if err != nil {
+				t.Fatalf("Unexpected error adding stream: %v", err)
+			}
+			defer mset.Delete()
+
+			var cons []*server.Consumer
+			for i := 0; i < 100; i++ {
+				o, err := mset.AddConsumer(&server.ConsumerConfig{
+					Durable:   fmt.Sprintf("d%d", i),
+					AckPolicy: server.AckExplicit,
+				})
+				if err != nil {
+					t.Fatalf("Expected no error, got %v", err)
+				}
+				defer o.Delete()
+				cons = append(cons, o)
+			}
+
+			wg := sync.WaitGroup{}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for _, c := range cons {
+					c.Delete()
+				}
+			}()
+			time.Sleep(10 * time.Millisecond)
+			s.Shutdown()
+		})
+	}
+}
+
 func TestJetStreamAddStreamMaxMsgSize(t *testing.T) {
 	cases := []struct {
 		name    string

